---
title: "Goat Plots"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
  pdf_document:
    toc: true
---

```{r setup, include=FALSE}
 knitr::opts_chunk$set(results='hide')
 library(tidyverse)
 library(gridExtra)
 set.seed(1987)
```

# Introduction
This document will walk through the PCA of the goats data used in the Genomic Bootcamp, doing the PCA within R and also using the PCA command in PLINK.

# Doing the Math in R

(The presumption is that the plink.exe file and the ADAPT files are in the working directory together, and prior to the running of this code, no changes have been made to the goat data files other than unzipping them and placing them into the directory.)

## Create Recoded File
```{r}
system("plink --bfile ADAPTmap_genotypeTOP_20160222_full --cow --autosome --geno 0.1 --mind 0.1 --maf 0.05 --nonfounders --allow-no-sex --recode --out ADAPTmap_TOP")
```

>48552 variants and 4532 cattle pass filters and QC.

## calculate Distances Between Individuals
```{r}
system("plink --cow --allow-no-sex --nonfounders --file ADAPTmap_TOP --distance-matrix --out GoatPCAInR")
```

>48552 variants and 4532 cattle pass filters and QC.

## Load data
```{r}
dist_populations<-read.table("GoatPCAInR.mdist",header=F)
```

### Extract breed names
```{r}
fam <- data.frame(famids=read.table("GoatPCAInR.mdist.id")[,1])
```

### Extract individual names 
```{r}
famInd <- data.frame(IID=read.table("GoatPCAInR.mdist.id")[,2])
```

## Perform PCA using the cmdscale function 
```{r}
mds_populations <- cmdscale(dist_populations,eig=T,5)
```

### Extract the eigen vectors
```{r}
eigenvec_populations <- cbind(fam,famInd,mds_populations$points)
```

### Proportion of variation captured by each eigen vector
```{r}
eigen_percent <- round(((mds_populations$eig)/sum(mds_populations$eig))*100,2)
```

## Visualize
```{r}
MathInRPlot<-ggplot(data = eigenvec_populations) +
  geom_point(mapping = aes(x = `1`, y = `2`, color=famids), show.legend = FALSE ) + 
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(title = "PCA Performed by R",
       x = paste0("Principal component 1 (",eigen_percent[1]," %)"),
       y = paste0("Principal component 2 (",eigen_percent[2]," %)"))
```

#Doing the Math in PLINK
In the example in the short video, the data set was filtered to be only a few goat breeds. I wanted to use the full data set to compare apples to apples, so did not filter any goat breeds. 
```{r}
system(str_c("plink --bfile ADAPTmap_genotypeTOP_20160222_full --chr-set 29 --autosome --nonfounders ",
             "--geno 0.1 --mind 0.1 --maf 0.05 ",
             "--make-bed --out afterQC"))
```

>48552 variants and 4532 samples pass filters and QC.
(Same as the other approach.)

## Perform PCA using PLINK 
```{r}
system("plink --bfile afterQC --chr-set 29 --pca --out GoatPCAInPLINK")
```

>48552 variants and 4532 samples pass filters and QC.

## Visualize
### Read in result files
```{r}
eigenValues <- read_delim("GoatPCAInPLINK.eigenval", delim = " ", col_names = F)
eigenVectors <- read_delim("GoatPCAInPLINK.eigenvec", delim = " ", col_names = F)
```

### Proportion of variation captured by each vector
```{r}
eigen_percent2 <- round((eigenValues / (sum(eigenValues))*100), 2)
```

### Plot
```{r}
MathInPLINKPlot <- ggplot(data = eigenVectors) +
  geom_point(mapping = aes(x = X3, y = X4, color = X1), show.legend = FALSE ) +
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(title = "PCA Performed by PLINK",
       x = paste0("Principal component 1 (",eigen_percent2[1,1]," %)"),
       y = paste0("Principal component 2 (",eigen_percent2[2,1]," %)"),
       colour = "Goat breeds", shape = "Goat breeds")
```

Legend was changed to FALSE since there are so many breeds due to not trimming down. Shape also removed, since there are not enough shapes when its not filtered.

# Plot Side by Side
```{r, include=TRUE}
combined_plots <- grid.arrange(MathInRPlot, MathInPLINKPlot, nrow = 1)
```

As you can see, the plots look very similar, though mirrored across the X axis. However, the percents for PC1 & PC2 in the axes labels are very different, despite being the same sample.

Notably, in trying to "backtrack this", when viewing the data frame `eigen_percent2` from which the percent is being called for the PLINK graph (lines 118 and 119 in the present document), the *fourth* value in this data frame happens to be 5.59 which is the Y axis percent on the *other* plot. However, the 8.2% from the X axis of that plot is not in `eigen_percent2`, so that may be a coincidence?